File: src/services/videoScanner.js
/**
 * 视频扫描服务
 */
const fs = require('fs');
const path = require('path');
const config = require('../config');
const videoCache = require('../models/videoCache');
const videoIdManager = require('../utils/videoId');

// 支持的视频扩展名
const VIDEO_EXTENSIONS = ['.mp4', '.mkv', '.avi', '.mov', '.webm'];

/**
 * 格式化文件大小
 */
function formatFileSize(bytes) {
  if (bytes < 1024) return bytes + ' B';
  if (bytes < 1024 * 1024) return (bytes / 1024).toFixed(2) + ' KB';
  if (bytes < 1024 * 1024 * 1024) return (bytes / (1024 * 1024)).toFixed(2) + ' MB';
  return (bytes / (1024 * 1024 * 1024)).toFixed(2) + ' GB';
}

/**
 * 格式化日期
 */
function formatDate(date) {
  return date.toLocaleString('zh-CN', {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit'
  });
}

/**
 * 异步递归扫描目录
 */
async function scanDirectoryRecursively(dirPath, dirName, dirConfigPath) {
  const videos = [];
  const visitedDirs = new Set();
  
  async function scan(currentPath, relativePath = '') {
    let realPath;
    try {
      realPath = await fs.promises.realpath(currentPath);
    } catch (err) {
      return; // 无法解析路径，跳过
    }
    if (visitedDirs.has(realPath)) return;
    visitedDirs.add(realPath);
    
    let items;
    try {
      items = await fs.promises.readdir(currentPath);
    } catch (err) {
      console.error(`无法读取目录 ${currentPath}: ${err.message}`);
      return;
    }
    
    for (const item of items) {
      if (item.startsWith('.')) continue;
      
      const itemPath = path.join(currentPath, item);
      const relativeItemPath = relativePath ? path.join(relativePath, item) : item;
      
      let stat;
      try {
        stat = await fs.promises.stat(itemPath);
      } catch (err) {
        continue;
      }
      
      if (stat.isDirectory()) {
        await scan(itemPath, relativeItemPath);
      } else if (stat.isFile()) {
        const ext = path.extname(item).toLowerCase();
        if (VIDEO_EXTENSIONS.includes(ext)) {
          videos.push({
            name: item,
            relativePath: relativeItemPath,
            fullPath: itemPath,
            dirName: dirName,
            size: formatFileSize(stat.size),
            sizeBytes: stat.size,
            modified: formatDate(stat.mtime),
            modifiedTime: stat.mtime.getTime()
          });
        }
      }
    }
  }
  
  await scan(dirPath);
  return videos;
}

/**
 * 扫描所有配置的目录
 */
async function scanAllDirectories() {
  // 检查是否已在扫描（通过文件锁）
  if (videoCache.isScanning()) {
    console.log('扫描已在进行中...');
    return videoCache.getVideos();
  }
  
  // 设置扫描锁
  await videoCache.setScanning(true);
  
  console.log('开始异步扫描视频目录...');
  
  const cfg = config.getConfig();
  const allVideos = [];
  
  try {
    for (const dirConfig of cfg.videoDirs) {
      const videoDir = path.resolve(dirConfig.path);
      
      let exists;
      try {
        await fs.promises.access(videoDir);
        exists = true;
      } catch {
        exists = false;
      }
      
      if (!exists) {
        console.warn(`警告: 目录不存在 - ${videoDir}`);
        continue;
      }
      
      try {
        const dirStat = await fs.promises.stat(videoDir);
        if (!dirStat.isDirectory()) {
          console.warn(`警告: 路径不是目录 - ${videoDir}`);
          continue;
        }
        
        console.log(`正在扫描目录: ${dirConfig.name}...`);
        const videos = await scanDirectoryRecursively(videoDir, dirConfig.name, dirConfig.path);
        allVideos.push(...videos);
        console.log(`✓ 扫描完成 ${dirConfig.name}: ${videos.length} 个视频`);
      } catch (err) {
        console.error(`错误: 扫描目录失败 ${dirConfig.path}: ${err.message}`);
      }
    }
    
    // 按名称排序
    allVideos.sort((a, b) => a.name.localeCompare(b.name));
    
    // 一次性写入缓存文件
    await videoCache.setVideos(allVideos);
    
    // 注册所有视频ID
    videoIdManager.registerVideos(allVideos);
    
    console.log(`=================================================`);
    console.log(`扫描完成！总计找到 ${allVideos.length} 个视频文件`);
    console.log(`=================================================`);
    
    return allVideos;
  } finally {
    // 无论成功失败，都释放扫描锁
    await videoCache.setScanning(false);
  }
}

module.exports = {
  scanDirectoryRecursively,
  scanAllDirectories,
  VIDEO_EXTENSIONS
};

File: go-impl/internal/service/videoScanner.go
// Package service 服务包
package service

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"

	"video-server/internal/config"
	"video-server/internal/model"
	"video-server/internal/utils"
)

// 支持的视频扩展名
var videoExtensions = map[string]bool{
	".mp4":  true,
	".mkv":  true,
	".avi":  true,
	".mov":  true,
	".webm": true,
}

// VideoScanner 视频扫描服务
type VideoScanner struct {
	cache *model.VideoCache
}

var (
	videoScanner *VideoScanner
	scannerOnce  sync.Once
)

// GetVideoScanner 获取视频扫描服务单例
func GetVideoScanner() *VideoScanner {
	scannerOnce.Do(func() {
		videoScanner = &VideoScanner{
			cache: model.GetVideoCache(),
		}
	})
	return videoScanner
}

// formatFileSize 格式化文件大小
func formatFileSize(bytes int64) string {
	if bytes < 1024 {
		return fmt.Sprintf("%d B", bytes)
	}
	if bytes < 1024*1024 {
		return fmt.Sprintf("%.2f KB", float64(bytes)/1024)
	}
	if bytes < 1024*1024*1024 {
		return fmt.Sprintf("%.2f MB", float64(bytes)/(1024*1024))
	}
	return fmt.Sprintf("%.2f GB", float64(bytes)/(1024*1024*1024))
}

// formatDate 格式化日期
func formatDate(t time.Time) string {
	return t.Format("2006/01/02 15:04")
}

// scanDirectoryRecursively 递归扫描目录
func (vs *VideoScanner) scanDirectoryRecursively(dirPath, dirName string) ([]model.Video, error) {
	var videos []model.Video
	visitedDirs := make(map[string]bool)
	
	var scan func(currentPath, relativePath string) error
	
	scan = func(currentPath, relativePath string) error {
		// 解析真实路径
		realPath, err := filepath.EvalSymlinks(currentPath)
		if err != nil {
			return nil // 无法解析路径，跳过
		}
		
		// 防止循环
		if visitedDirs[realPath] {
			return nil
		}
		visitedDirs[realPath] = true
		
		// 读取目录
		entries, err := os.ReadDir(currentPath)
		if err != nil {
			return err
		}
		
		for _, entry := range entries {
			// 跳过隐藏文件
			if strings.HasPrefix(entry.Name(), ".") {
				continue
			}
			
			itemPath := filepath.Join(currentPath, entry.Name())
			var relItemPath string
			if relativePath != "" {
				relItemPath = filepath.Join(relativePath, entry.Name())
			} else {
				relItemPath = entry.Name()
			}
			
			info, err := entry.Info()
			if err != nil {
				continue
			}
			
			if entry.IsDir() {
				scan(itemPath, relItemPath)
			} else {
				ext := strings.ToLower(filepath.Ext(entry.Name()))
				if videoExtensions[ext] {
					videos = append(videos, model.Video{
						Name:         entry.Name(),
						RelativePath: relItemPath,
						FullPath:     itemPath,
						DirName:      dirName,
						Size:         formatFileSize(info.Size()),
						SizeBytes:    info.Size(),
						Modified:     formatDate(info.ModTime()),
						ModifiedTime: info.ModTime().UnixMilli(),
					})
				}
			}
		}
		
		return nil
	}
	
	err := scan(dirPath, "")
	return videos, err
}

// ScanAllDirectories 扫描所有配置的目录
func (vs *VideoScanner) ScanAllDirectories() ([]model.Video, error) {
	// 检查是否已在扫描
	if vs.cache.IsScanning() {
		fmt.Println("扫描已在进行中...")
		return vs.cache.GetVideos(), nil
	}
	
	// 设置扫描锁
	vs.cache.SetScanning(true)
	defer vs.cache.SetScanning(false)
	
	fmt.Println("开始异步扫描视频目录...")
	
	cfg := config.GetConfig()
	var allVideos []model.Video
	
	for _, dirConfig := range cfg.VideoDirs {
		videoDir, err := filepath.Abs(dirConfig.Path)
		if err != nil {
			fmt.Printf("警告: 无法解析路径 - %s\n", dirConfig.Path)
			continue
		}
		
		// 检查目录是否存在
		if _, err := os.Stat(videoDir); os.IsNotExist(err) {
			fmt.Printf("警告: 目录不存在 - %s\n", videoDir)
			continue
		}
		
		// 检查是否是目录
		info, err := os.Stat(videoDir)
		if err != nil || !info.IsDir() {
			fmt.Printf("警告: 路径不是目录 - %s\n", videoDir)
			continue
		}
		
		fmt.Printf("正在扫描目录: %s...\n", dirConfig.Name)
		videos, err := vs.scanDirectoryRecursively(videoDir, dirConfig.Name)
		if err != nil {
			fmt.Printf("错误: 扫描目录失败 %s: %v\n", dirConfig.Path, err)
			continue
		}
		
		allVideos = append(allVideos, videos...)
		fmt.Printf("✓ 扫描完成 %s: %d 个视频\n", dirConfig.Name, len(videos))
	}
	
	// 按名称排序
	sort.Slice(allVideos, func(i, j int) bool {
		return allVideos[i].Name < allVideos[j].Name
	})

	// 注册视频ID（会设置 video.ID）
	utils.GetVideoIDManager().RegisterVideos(allVideos)

	// 保存到缓存
	vs.cache.SetVideos(allVideos)
	
	fmt.Println("=================================================")
	fmt.Printf("扫描完成！总计找到 %d 个视频文件\n", len(allVideos))
	fmt.Println("=================================================")
	
	return allVideos, nil
}
